File: SnehaS/ads/rbtree.c
#include <stdio.h>
#include <stdlib.h>

#define RED 1
#define BLACK 0

struct Node {
    int data, color;
    struct Node *left, *right, *parent;
};

struct Node *root, *NIL;
struct Node* createNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data;
    node->color = RED;
    node->left = node->right = NIL;
    node->parent = NULL;
    return node;
}

void leftRotate(struct Node *x) {
    struct Node *y = x->right;
    x->right = y->left;

    if (y->left != NIL)
        y->left->parent = x;

    y->parent = x->parent;

    if (x->parent == NULL)
        root = y;
    else if (x == x->parent->left)
        x->parent->left = y;
    else
        x->parent->right = y;

    y->left = x;
    x->parent = y;
}

void rightRotate(struct Node *y) {
    struct Node *x = y->left;
    y->left = x->right;

    if (x->right != NIL)
        x->right->parent = y;

    x->parent = y->parent;

    if (y->parent == NULL)
        root = x;
    else if (y == y->parent->right)
        y->parent->right = x;
    else
        y->parent->left = x;

    x->right = y;
    y->parent = x;
}

void fixInsert(struct Node *k) {
    while (k->parent && k->parent->color == RED) {
        if (k->parent == k->parent->parent->left) {
            struct Node *u = k->parent->parent->right;

            if (u->color == RED) {
                k->parent->color = BLACK;
                u->color = BLACK;
                k->parent->parent->color = RED;
                k = k->parent->parent;
            } else {
                if (k == k->parent->right) {
                    k = k->parent;
                    leftRotate(k);
                }
                k->parent->color = BLACK;
                k->parent->parent->color = RED;
                rightRotate(k->parent->parent);
            }
        } else {
            struct Node *u = k->parent->parent->left;

            if (u->color == RED) {
                k->parent->color = BLACK;
                u->color = BLACK;
                k->parent->parent->color = RED;
                k = k->parent->parent;
            } else {
                if (k == k->parent->left) {
                    k = k->parent;
                    rightRotate(k);
                }
                k->parent->color = BLACK;
                k->parent->parent->color = RED;
                leftRotate(k->parent->parent);
            }
        }
    }
    root->color = BLACK;
}

void insert(int data) {
    struct Node *node = createNode(data);
    struct Node *y = NULL;
    struct Node *x = root;

    while (x != NIL) {
        y = x;
        if (data < x->data)
            x = x->left;
        else
            x = x->right;
    }

    node->parent = y;

    if (y == NULL)
        root = node;
    else if (data < y->data)
        y->left = node;
    else
        y->right = node;

    if (node->parent == NULL) {
        node->color = BLACK;
        return;
    }

    if (node->parent->parent == NULL)
        return;

    fixInsert(node);
}

struct Node* search(struct Node* node, int key) {
    if (node == NIL || key == node->data)
        return node;
    if (key < node->data)
        return search(node->left, key);
    return search(node->right, key);
}

struct Node* minimum(struct Node* node) {
    while (node->left != NIL)
        node = node->left;
    return node;
}

void transplant(struct Node *u, struct Node *v) {
    if (u->parent == NULL)
        root = v;
    else if (u == u->parent->left)
        u->parent->left = v;
    else
        u->parent->right = v;

    v->parent = u->parent;
}

void fixDelete(struct Node *x) {
    while (x != root && x->color == BLACK) {
        if (x == x->parent->left) {
            struct Node *s = x->parent->right;

            if (s->color == RED) {
                s->color = BLACK;
                x->parent->color = RED;
                leftRotate(x->parent);
                s = x->parent->right;
            }

            if (s->left->color == BLACK && s->right->color == BLACK) {
                s->color = RED;
                x = x->parent;
            } else {
                if (s->right->color == BLACK) {
                    s->left->color = BLACK;
                    s->color = RED;
                    rightRotate(s);
                    s = x->parent->right;
                }

                s->color = x->parent->color;
                x->parent->color = BLACK;
                s->right->color = BLACK;
                leftRotate(x->parent);
                x = root;
            }
        } else {
            struct Node *s = x->parent->left;

            if (s->color == RED) {
                s->color = BLACK;
                x->parent->color = RED;
                rightRotate(x->parent);
                s = x->parent->left;
            }

            if (s->right->color == BLACK && s->left->color == BLACK) {
                s->color = RED;
                x = x->parent;
            } else {
                if (s->left->color == BLACK) {
                    s->right->color = BLACK;
                    s->color = RED;
                    leftRotate(s);
                    s = x->parent->left;
                }

                s->color = x->parent->color;
                x->parent->color = BLACK;
                s->left->color = BLACK;
                rightRotate(x->parent);
                x = root;
            }
        }
    }
    x->color = BLACK;
}

void deleteValue(int data) {
    struct Node *z = search(root, data);
    if (z == NIL) {
        printf("Value not found\n");
        return;
    }

    struct Node *y = z;
    struct Node *x;
    int y_original_color = y->color;

    if (z->left == NIL) {
        x = z->right;
        transplant(z, z->right);
    } else if (z->right == NIL) {
        x = z->left;
        transplant(z, z->left);
    } else {
        y = minimum(z->right);
        y_original_color = y->color;
        x = y->right;

        if (y->parent == z)
            x->parent = y;
        else {
            transplant(y, y->right);
            y->right = z->right;
            y->right->parent = y;
        }

        transplant(z, y);
        y->left = z->left;
        y->left->parent = y;
        y->color = z->color;
    }

    if (y_original_color == BLACK)
        fixDelete(x);
}

void inorder(struct Node *node) {
    if (node != NIL) {
        inorder(node->left);
        printf("%d(%s) ", node->data, node->color == RED ? "R" : "B");
        inorder(node->right);
    }
}

int main() {
    NIL = (struct Node*)malloc(sizeof(struct Node));
    NIL->color = BLACK;
    NIL->left = NIL->right = NIL;

    root = NIL;

    int choice, value;

    while (1) {
        printf("\n1.Insert\n2.Delete\n3.Display\n4.Exit\n");
        printf("Enter choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter value: ");
                scanf("%d", &value);
                insert(value);
                break;

            case 2:
                printf("Enter value: ");
                scanf("%d", &value);
                deleteValue(value);
                break;

            case 3:
                printf("Inorder: ");
                inorder(root);
                printf("\n");
                break;

            case 4:
                exit(0);

            default:
                printf("Invalid choice\n");
        }
    }
}

File: Nishan/rbtree.py
class Node:
 def __init__(self, value, color=1):
  self.value = value
  self.color = color
  self.left = self.right = self.parent = None

class RedBlackTree:
 def __init__(self):
  self.NIL = Node(0, 0)
  self.root = self.NIL

 def left_rotate(self, node):
  right_child = node.right
  node.right = right_child.left
  if right_child.left != self.NIL:
   right_child.left.parent = node
  right_child.parent = node.parent
  if node.parent is None:
   self.root = right_child
  elif node == node.parent.left:
   node.parent.left = right_child
  else:
   node.parent.right = right_child
  right_child.left = node
  node.parent = right_child

 def right_rotate(self, node):
  left_child = node.left
  node.left = left_child.right
  if left_child.right != self.NIL:
   left_child.right.parent = node
  left_child.parent = node.parent
  if node.parent is None:
   self.root = left_child
  elif node == node.parent.right:
   node.parent.right = left_child
  else:
   node.parent.left = left_child
  left_child.right = node
  node.parent = left_child

 def insert(self, value):
  new_node = Node(value)
  new_node.left = new_node.right = self.NIL
  parent = None
  current = self.root
  while current != self.NIL:
   parent = current
   if value < current.value:
    current = current.left
   else:
    current = current.right
  new_node.parent = parent
  if parent is None:
   self.root = new_node
  elif value < parent.value:
   parent.left = new_node
  else:
   parent.right = new_node
  if new_node.parent is None:
   new_node.color = 0
   return
  if new_node.parent.parent is None:
   return
  self.fix_insert(new_node)

 def fix_insert(self, node):
  while node.parent.color == 1:
   if node.parent == node.parent.parent.right:
    uncle = node.parent.parent.left
    if uncle.color == 1:
     uncle.color = 0
     node.parent.color = 0
     node.parent.parent.color = 1
     node = node.parent.parent
    else:
     if node == node.parent.left:
      node = node.parent
      self.right_rotate(node)
     node.parent.color = 0
     node.parent.parent.color = 1
     self.left_rotate(node.parent.parent)
   else:
    uncle = node.parent.parent.right
    if uncle.color == 1:
     uncle.color = 0
     node.parent.color = 0
     node.parent.parent.color = 1
     node = node.parent.parent
    else:
     if node == node.parent.right:
      node = node.parent
      self.left_rotate(node)
     node.parent.color = 0
     node.parent.parent.color = 1
     self.right_rotate(node.parent.parent)
   if node == self.root:
    break
  self.root.color = 0

 def transplant(self, u, v):
  if u.parent is None:
   self.root = v
  elif u == u.parent.left:
   u.parent.left = v
  else:
   u.parent.right = v
  v.parent = u.parent

 def find_min(self, node):
  while node.left != self.NIL:
   node = node.left
  return node

 def delete(self, value):
  node = self.root
  while node != self.NIL and node.value != value:
   node = node.left if value < node.value else node.right
  if node == self.NIL:
   print("Value not found in tree")
   return
  temp = node
  original_color = temp.color
  if node.left == self.NIL:
   fix_node = node.right
   self.transplant(node, node.right)
  elif node.right == self.NIL:
   fix_node = node.left
   self.transplant(node, node.left)
  else:
   temp = self.find_min(node.right)
   original_color = temp.color
   fix_node = temp.right
   if temp.parent != node:
    self.transplant(temp, temp.right)
    temp.right = node.right
    temp.right.parent = temp
   self.transplant(node, temp)
   temp.left = node.left
   temp.left.parent = temp
   temp.color = node.color
  if original_color == 0:
   self.fix_delete(fix_node)

 def fix_delete(self, node):
  while node != self.root and node.color == 0:
   if node == node.parent.left:
    sibling = node.parent.right
    if sibling.color == 1:
     sibling.color = 0
     node.parent.color = 1
     self.left_rotate(node.parent)
     sibling = node.parent.right
    if sibling.left.color == 0 and sibling.right.color == 0:
     sibling.color = 1
     node = node.parent
    else:
     if sibling.right.color == 0:
      sibling.left.color = 0
      sibling.color = 1
      self.right_rotate(sibling)
      sibling = node.parent.right
     sibling.color = node.parent.color
     node.parent.color = 0
     sibling.right.color = 0
     self.left_rotate(node.parent)
     node = self.root
   else:
    sibling = node.parent.left
    if sibling.color == 1:
     sibling.color = 0
     node.parent.color = 1
     self.right_rotate(node.parent)
     sibling = node.parent.left
    if sibling.left.color == 0 and sibling.right.color == 0:
     sibling.color = 1
     node = node.parent
    else:
     if sibling.left.color == 0:
      sibling.right.color = 0
      sibling.color = 1
      self.left_rotate(sibling)
      sibling = node.parent.left
     sibling.color = node.parent.color
     node.parent.color = 0
     sibling.left.color = 0
     self.right_rotate(node.parent)
     node = self.root
  node.color = 0

 def display(self, node=None, indent="", last=True):
  if node is None:
   node = self.root
  if node != self.NIL:
   print(indent + ("R----" if last else "L----") + f"{node.value} ({'RED' if node.color else 'BLACK'})")
   indent += "     " if last else "|    "
   self.display(node.left, indent, False)
   self.display(node.right, indent, True)

tree = RedBlackTree()
values = list(map(int, input("Enter values: ").split()))
for val in values:
 tree.insert(val)

print("\nTree before deletion:")
tree.display()

delete_val = int(input("\nEnter value to delete: "))
tree.delete(delete_val)

print("\nTree after deletion:")
tree.display()
